src/boot/chid.h

```diff
@@ -36,6 +39,14 @@ typedef struct Device {
                         uint32_t name_offset;          /* nul-terminated string or 0 if not present */
                         uint32_t compatible_offset;    /* nul-terminated string or 0 if not present */
                 } devicetree;
+                struct {
```

Comment:
@poettering 
> moreover, i think when we actually start adding more device types we should probably change the parser to actually iterate through the thing always taking the specified structure size into account, instead of assuming all structures are the same length

The only way that these structures can be of different lengths in the future is if the `Device` would be split into a few separate structures:
```C
typedef struct DeviceHeader {
        uint32_t descriptor; /* The highest four bit encode the type of entry, the other 28 bit encode the
                              * size of the structure. Use the macros above to generate or take apart this
                              * field. */
        EFI_GUID chid;
} _packed_ DeviceHeader;

typedef struct DeviceDeviceTree {
        DeviceHeader hdr;
        /* These offsets are relative to the beginning of the .hwids PE section. */
        uint32_t name_offset;          /* nul-terminated string or 0 if not present */
        uint32_t compatible_offset;    /* nul-terminated string or 0 if not present */
} _packed_ DeviceDeviceTree;

typedef struct DeviceEfiFirmware {
        DeviceHeader hdr;
        /* Offsets are relative to the beginning of the .hwids PE section.
             They are nul-terminated strings when present or 0 if not present */
        uint32_t id_offset;         /* identifier for the firmware blob */
        uint32_t metadata_offset;   /* firmware metadata string */
        uint32_t compatible_offset; /* compatibility identifier to match a specific fw blob */
} _packed_ DeviceEfiFirmware;
```

`DeviceDeviceTree` structure definition is compatible with current `Device` definition, so it seems like a better approach. At least `ukify` won't require any additional changes for currently supported types (`DEVICE_TYPE_DEVICETREE` as of now) each time a new type is added.

Reply:
yes, this is the way to go, if the struct needs diffrent space for the types (and i think it does)